rmf_fleet_adapter/src/rmf_fleet_adapter/agv/FleetUpdateHandle.cpp

```diff
@@ -557,63 +492,124 @@ void FleetUpdateHandle::Implementation::bid_notice_cb(
 void FleetUpdateHandle::Implementation::dispatch_request_cb(
   const DispatchRequest::SharedPtr msg)
 {
   if (msg->fleet_name != name)
     return;
 
   const std::string id = msg->task_profile.task_id;
-  const auto task_it = bid_notice_assignments.find(id);
 
-  if (task_it == bid_notice_assignments.end())
-    return;
+  if (msg->method == DispatchRequest::ADD)
+  {
+    const auto task_it = bid_notice_assignments.find(id);
+    if (task_it == bid_notice_assignments.end())
+      return;
 
-  RCLCPP_INFO(
-    node->get_logger(),
-    "Bid for task_id:[%s] awarded to fleet [%s]",
-    id.c_str(), name.c_str());
+    RCLCPP_INFO(
+      node->get_logger(),
+      "Bid for task_id:[%s] awarded to fleet [%s]",
+      id.c_str(), name.c_str());
 
-  // We currently only support adding tasks
-  if (msg->method != DispatchRequest::ADD)
-    return;
-  
-  const auto& assignments = task_it->second;
-  
-  if (assignments.size() != task_managers.size())
-  {
-    RCLCPP_ERROR(
+    auto& assignments = task_it->second;
+    
+    if (assignments.size() != task_managers.size())
+    {
+      RCLCPP_ERROR(
+        node->get_logger(),
+        "The number of available robots does not match that in the assignments "
+        "for task_id:[%s]. This request will be ignored.", id.c_str());
+
+      return;
+    }
+
+    // Here we make sure none of the tasks in the assignments has already begun
+    // execution. If so, we replan assignments until a valid set is obtained 
+    // and only then update the task manager queues
+    bool valid_assignments = false;
+    while (!valid_assignments)
+    {
+      valid_assignments = is_valid_assignments(assignments);
+      if (!valid_assignments)
+      {
+        const auto request_it = generated_requests.find(id);
+        if (request_it == generated_requests.end())
+          return;
+        const auto replan_results = allocate_tasks(request_it->second);
+        if (!replan_results)
+        {
+          RCLCPP_WARN(
+            node->get_logger(),
+            "Unable to replan assignments for task_id:[%s]", id.c_str());
+          return;
+        }
+        assignments = replan_results.value();
+        // We re-check the assignments as a task could have started while replanning
+        valid_assignments = is_valid_assignments(assignments);
+      }
+    }
+
+    std::size_t index = 0;
+    for (auto& t : task_managers)
+    {
+      t.second->set_queue(assignments[index]);
+      ++index;
+    }
+
+    current_assignment_cost = task_planner->compute_cost(assignments);
+
+    RCLCPP_INFO(
       node->get_logger(),
-      "The number of available robots do not match that in the assignments "
-      "for task_id:[%s]. This request will be ignored", id.c_str());
+      "Assignments updated for robots in fleet [%s] to accommodate task_id:[%s]",
+      name.c_str(), id.c_str());
+  }
 
+  else if (msg->method == DispatchRequest::CANCEL)
+  {
+    // TODO(YV)
+  }
+
+  else
+  {
     return;
```

Comment:
We should probably return a warning message saying that we received an unsupported dispatch request type.

I would also suggest that we put a "Not implemented yet" warning message for the `CANCEL` case above.

Reply:
Done https://github.com/osrf/rmf_core/pull/265/commits/347e4bff842986b6ba7dc0d653bc7743f53b57b2